README.md
# Chess-Project
Chess with basic functionality

The goal of this project is to build a minimally working chess game using pyglet.
That is, the users can click to play the game until checkmate; and they'd have to reopen the window to play again.
Optimization is not a priority, only for fun.

The folowing lines list the possible issues to be created and fixed until the end of the project.
-handle piece capturing
-handle en passant
-handle pawn promotion (may require extra work if to allow underpromotion)
-given a position, determine whether a king is on check
-given a position, determine whether a king is on checkmate

Note that this wouldn't be a windows executable file. I've tried converting from .py to .exe but it didn't work.


changes.txt
21/08/2020
Pieces are now initialized using square and not file, rank
All pieces now move and capture normally, except captured pieces are to be eliminated from the board
En passant not yet implemented
Next:
Eliminate captured pieces
Implement en passant

20/08/2020
corrected bottommost spelling
new function piece_at_square for chessgame.py
bishops and rooks has incorrect implementation of blockable movement
finally fixed that implementation; consequentially also applies to queens
Next: use square for __init__ of pieces instead of file, rank

Before 20/08/2020
-interactive point and click chess
-pieces can move their normal way if not for blockings


chessproject/utils.py
import pyglet
import resources
#import abc

DEFAULT_PIECE_SCALE = 0.8

class Piece():
    def __init__(self, file, rank, player):
        self.player = player
        self.type = type
        self.file = file
        self.rank = rank
        self.letter = "P"

    def __str__(self):
        return self.letter + chr(self.file+97) + str(self.rank+1)

    def get_sprite(self):
        image = resources.all_pieces[self.player][self.letter]
        resources.center_image(image)
        sprite = pyglet.sprite.Sprite(image)
        sprite.scale *= DEFAULT_PIECE_SCALE

        sprite.x = 63 + (self.file + 0.5)*(59.25)       # we don't want magic numbers
        sprite.y = 62 + (self.rank + 0.5)*(60)          # we don't want magic numbers

        return sprite

    #@abc.abstractmethod
    def move(self, square):
        # if self.guards(square):
        # file = x rank = y
        pass

    #@abc.abstractmethod
    def guards(self, square):
        return False


class King(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "K"
        self.sprite = self.get_sprite()

class Queen(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "Q"
        self.sprite = self.get_sprite()

class Bishop(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "B"
        self.sprite = self.get_sprite()

class Knight(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "N"
        self.sprite = self.get_sprite()

class Rook(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "R"
        self.sprite = self.get_sprite()

class Pawn(Piece):
    def __init__(self, file, rank, player):
        super().__init__(file, rank, player)
        self.letter = "P"
        self.sprite = self.get_sprite()


def get_pieces(position_str):
    pieces = []
    rows = position_str.split('/')
    i = 0
    while (i < 8):
        j = 0
        while (j < 8):
            char = rows[i][j]
            is_int = True
            try:
                int(char)
            except ValueError:
                is_int = False
            if not is_int:
                pieces.append(create_piece(char, j, 7-i))
                j += 1
            else:
                j += int(char)
        i += 1
    return pieces


def create_piece(letter, file, rank):
    piece = letter.upper()
    player = 0 if piece == letter else 1
    if piece == "K":
        return King(file, rank, player)
    if piece == "Q":
        return Queen(file, rank, player)
    if piece == "B":
        return Bishop(file, rank, player)
    if piece == "N":
        return Knight(file, rank, player)
    if piece == "R":
        return Rook(file, rank, player)
    if piece == "P":
        return Pawn(file, rank, player)



def file_rank_of(square):
    return (ord(square[0])-97, int(square[1]))

if __name__ == '__main__':
    k = King(7, 0)
    print(k)


chessproject/chessgame.py
import pyglet
import resources
import board_and_pieces as bp


class ChessGame:
    initial_position = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
    def __init__(self, position_str, current_turn, castling_posibilities, en_passant_square):
        self.board = bp.Board()
        self.pieces = self.get_pieces(position_str)
        self.current_turn = current_turn
        self.castling_posibilities = castling_posibilities
        self.en_passant_square= en_passant_square

        self.batch = pyglet.graphics.Batch()
        for piece in self.pieces:
            piece.sprite.batch = self.batch

        self.selected_piece = None

    def __str__(self):
        return [str(piece) for piece in self.pieces]

    def get_pieces(self, position_str):
        pieces = []
        rows = position_str.split('/')
        i = 0
        while (i < 8):
            j = 0
            while (j < 8):
                char = rows[i][j]
                is_int = True
                try:
                    int(char)
                except ValueError:
                    is_int = False
                if not is_int:
                    sq = self.board.square_of_fr(j, 7-i)
                    pieces.append(self.create_piece(char, sq))
                    j += 1
                else:
                    j += int(char)
            i += 1
        return pieces

    def is_out_of_bounds_fr(self, file, rank):
        if file > 7 or rank > 7 or file < 0 or rank < 0:
            return True


    def create_piece(self, letter, square):
        piece = letter.upper()
        player = 0 if piece == letter else 1
        if piece == "K":
            return bp.King(square, player, self)
        if piece == "Q":
            return bp.Queen(square, player, self)
        if piece == "B":
            return bp.Bishop(square, player, self)
        if piece == "N":
            return bp.Knight(square, player, self)
        if piece == "R":
            return bp.Rook(square, player, self)
        if piece == "P":
            return bp.Pawn(square, player, self)

    def on_draw(self):
        self.board.draw()
        self.batch.draw()

    def on_click(self, x, y):
        file, rank = self.board.file_rank_of_xy(x, y)
        if self.is_out_of_bounds_fr(file, rank):
            pass
        elif self.selected_piece is not None:
            self.selected_piece.on_click(x, y)
            self.selected_piece = None
        else:
            sq = self.board.square_of_xy(x, y)
            piece = self.piece_at_square(sq)
            if not piece.is_my_turn():
                return False
            print(str(piece) + " is selected")
            self.selected_piece = piece
            piece.on_click(x, y)

    def piece_at_square(self, square):
        # can be O(1) with 64 square objects
        for piece in self.pieces:
            #print(square, piece.square)
            if square == piece.square:
                return piece
        return None

    def pawn_of_ep_square(self):
        if self.en_passant_square is None: return None
        file, rank = self.board.file_rank_of_sq(self.en_passant_square)
        forward = None
        if rank==2:
            forward = 1
        if rank==5:
            forward = -1
        pawn_at_sq = self.board.square_of_fr(file, rank+forward)
        return pawn_at_sq if isinstance(pawn_at_sq, bp.Pawn) else None

    def finish_turn(self):
        self.current_turn += 1

    def remove_piece_at_sq(self, square):
        self.pieces.remove(self.piece_at_square(square))

    def remove_piece(self, piece):
        self.pieces.remove(piece)


chessproject/resources.py
import pyglet

def center_image(image):
    """Sets an image's anchor point to its center"""
    image.anchor_x = image.width//2
    image.anchor_y = image.height//2


board_sprite = pyglet.resource.image('res/board.png')
board_sprite = pyglet.sprite.Sprite(board_sprite)


wking = pyglet.resource.image('res/wking.png')
wqueen = pyglet.resource.image('res/wqueen.png')
wbishop = pyglet.resource.image('res/wbishop.png')
wknight = pyglet.resource.image('res/wknight.png')
wrook = pyglet.resource.image('res/wrook.png')
wpawn = pyglet.resource.image('res/wpawn.png')

bking = pyglet.resource.image('res/bking.png')
bqueen = pyglet.resource.image('res/bqueen.png')
bbishop = pyglet.resource.image('res/bbishop.png')
bknight = pyglet.resource.image('res/bknight.png')
brook = pyglet.resource.image('res/brook.png')
bpawn = pyglet.resource.image('res/bpawn.png')

all_pieces = [{
    "K": wking,
    "Q": wqueen,
    "B": wbishop,
    "N": wknight,
    "R": wrook,
    "P": wpawn
},
{
    "K": bking,
    "Q": bqueen,
    "B": bbishop,
    "N": bknight,
    "R": brook,
    "P": bpawn
}]


chessproject/board_and_pieces.py
import pyglet
from math import copysign
import resources
import abc

DEFAULT_PIECE_SCALE = 0.8
DEFAULT_BOARD_SCALE = 1
BOARD_LEFTMOST = 63*DEFAULT_BOARD_SCALE
BOARD_BOTTOMMOST = 62*DEFAULT_BOARD_SCALE
BOARD_SQUARE_WIDTH = 59.25*DEFAULT_BOARD_SCALE
BOARD_SQUARE_HEIGHT = 60*DEFAULT_BOARD_SCALE


class Board:
    def __init__(self):
        self.sprite = resources.board_sprite
        self.leftmost_x = BOARD_LEFTMOST
        self.bottommost_y = BOARD_BOTTOMMOST
        self.square_width = BOARD_SQUARE_WIDTH
        self.square_height = BOARD_SQUARE_HEIGHT

    def draw(self):
        self.sprite.draw()

    def file_rank_of_sq(self, square):
        return (ord(square[0])-97, int(square[1])-1)

    def square_of_fr(self, file, rank):
        return chr(file+97) + str(rank+1)

    def file_rank_of_xy(self, tx, ty):
        file = round((tx-self.leftmost_x)/self.square_width - 0.5)
        rank = round((ty-self.bottommost_y)/self.square_height - 0.5)
        return (file, rank)

    def square_of_xy(self, tx, ty):
        f, r = self.file_rank_of_xy(tx, ty)
        return self.square_of_fr(f, r)




class Piece():
    def __init__(self, square, player, chessgame):
        self.chessgame = chessgame
        self.board = chessgame.board
        self.player = player
        self.type = type
        self.letter = "P"

        file, rank = self.board.file_rank_of_sq(square)
        self.square = square
        self.file = file
        self.rank = rank


        self.is_selected = False

    def __str__(self):
        return self.letter + chr(self.file+97) + str(self.rank+1)

    def get_sprite(self):
        image = resources.all_pieces[self.player][self.letter]
        resources.center_image(image)
        sprite = pyglet.sprite.Sprite(image)
        sprite.scale *= DEFAULT_PIECE_SCALE
        sprite.x = self.board.leftmost_x + (self.file + 0.5)*self.board.square_width
        sprite.y = self.board.bottommost_y + (self.rank + 0.5)*self.board.square_height
        return sprite

    def update_sprite_position(self):
        self.sprite.x = self.board.leftmost_x + (self.file + 0.5)*self.board.square_width
        self.sprite.y = self.board.bottommost_y + (self.rank + 0.5)*self.board.square_height

    def square_has_enemy(self, square):
        piece = self.chessgame.piece_at_square(square)
        if piece is None: return False
        return piece.player != self.player

    def square_has_friendly(self, square):
        piece = self.chessgame.piece_at_square(square)
        if piece is None: return False
        return piece.player == self.player

    def is_on(self, tx, ty):
        return abs(self.sprite.x - tx) < self.sprite.width//2 and abs(self.sprite.y - ty) < self.sprite.height//2

    def on_click(self, tx, ty):
        if self.is_on(tx, ty):
            if not self.is_selected:
                self.is_selected = True
            else:
                self.is_selected = False
        else:
            if self.is_selected:
                self.move(self.chessgame.board.square_of_xy(tx, ty))
                self.is_selected = False

    def is_my_turn(self):
        return self.player == self.chessgame.current_turn%2

    def move(self, square):
        is_capturing = self.square_has_enemy(square)
        if self.is_legal_move(square):
            if is_capturing:
                self.chessgame.remove_piece_at_sq(square)

            file, rank = self.chessgame.board.file_rank_of_sq(square)
            self.file = file
            self.rank = rank
            self.square = square
            self.update_sprite_position()

            self.chessgame.finish_turn()

    def guards(self, square):
        return False

    def is_legal_move(self, square):
        return self.guards(square) and not self.square_has_friendly(square)


class King(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "K"
        self.sprite = self.get_sprite()

    def guards(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank - self.rank

        return (abs(dr) == 1 or abs(df) == 1) and abs(dr) < 2 and abs(df) < 2



class Queen(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "Q"
        self.sprite = self.get_sprite()

    def guards(self, square):
        return Rook.guards(self, square) or Bishop.guards(self, square)

class Bishop(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "B"
        self.sprite = self.get_sprite()

    def guards(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank - self.rank

        is_basic_attack = abs(df) == abs(dr)
        is_blocked = False
        if not is_basic_attack: return False

        stepf = int(copysign(1, df))
        stepr = int(copysign(1, dr))
        f = self.file
        r = self.rank

        for i in range(max(abs(df), abs(dr))):
            if i == 0: continue
            sq = self.board.square_of_fr(f + stepf*i, r + stepr*i)
            if self.chessgame.piece_at_square(sq) is not None:
                is_blocked = True
                break

        return is_basic_attack and not is_blocked

class Knight(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "N"
        self.sprite = self.get_sprite()

    def guards(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank -self.rank

        return (abs(df) == 2 and abs(dr) == 1) or (abs(dr) == 2 and abs(df) == 1)

class Rook(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "R"
        self.sprite = self.get_sprite()

    def guards(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank - self.rank

        is_basic_attack = (df == 0 and dr != 0) or (dr == 0 and df != 0)
        is_blocked = False
        if not is_basic_attack: return False

        stepf = int(copysign(1, df)) if df else 0
        stepr = int(copysign(1, dr)) if dr else 0
        f = self.file
        r = self.rank

        for i in range(max(abs(df), abs(dr))):
            if i == 0: continue
            sq = self.board.square_of_fr(f + stepf*i, r + stepr*i)
            if self.chessgame.piece_at_square(sq) is not None:
                is_blocked = True
                break

        return is_basic_attack and not is_blocked

class Pawn(Piece):
    def __init__(self, square, player, chessgame):
        super().__init__(square, player, chessgame)
        self.letter = "P"
        self.sprite = self.get_sprite()

        file, rank = self.board.file_rank_of_sq(square)
        self.initial_rank = 1 if self.player == 0 else 6
        self.forward = 1 if self.player == 0 else -1
        self.has_moved = False if rank==self.initial_rank else True

        self.ep_square = self.board.square_of_fr(file, rank + self.forward)
        self.ep_square_is_active = False
        self.ep_square_is_active_turn = None

    def is_initial_move(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        r0 = self.rank
        return not self.has_moved and file==self.file and ((self.player==0 and rank==r0+2) or (self.player==1 and rank==r0-2))

    def can_advance(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank - self.rank
        is_white = self.player == 0
        is_black = self.player == 1
        stepr = int(copysign(1, dr))

        if df != 0: return False
        on_file = df == 0
        is_forward = (dr > 0 and is_white) or (dr < 0 and self.player == is_black)
        is_initial = self.is_initial_move(square)
        is_initial_not_blocked = is_initial and self.chessgame.piece_at_square(self.board.square_of_fr(self.file, self.rank + stepr)) is None
        is_empty = self.chessgame.piece_at_square(square) is None
        return on_file and is_forward and is_empty and (is_initial_not_blocked or abs(dr) == 1)

    def guards(self, square):
        file, rank = self.board.file_rank_of_sq(square)
        df = file - self.file
        dr = rank -self.rank

        is_diag = abs(df) == 1 and abs(dr) == 1
        return is_diag and ((self.player == 0 and dr > 0) or (self.player == 1 and dr < 0))

    def is_legal_move(self, square):
        is_capturing = self.guards(square) and self.square_has_enemy(square)
        is_advancing = self.can_advance(square)
        return is_capturing or is_advancing
